# Project 1: Trading with Momentum
## Instructions
Each problem consists of a function to implement and instructions on how to implement the function.  The parts of the function that need to be implemented are marked with a `# TODO` comment. After implementing the function, run the cell to test it against the unit tests we've provided. For each problem, we provide one or more unit tests from our `project_tests` package. These unit tests won't tell you if your answer is correct, but will warn you of any major errors. Your code will be checked for the correct solution when you submit it to Udacity.

## Packages
When you implement the functions, you'll only need to you use the packages you've used in the classroom, like [Pandas](https://pandas.pydata.org/) and [Numpy](http://www.numpy.org/). These packages will be imported for you. We recommend you don't add any import statements, otherwise the grader might not be able to run your code.

The other packages that we're importing are `helper`, `project_helper`, and `project_tests`. These are custom packages built to help you solve the problems.  The `helper` and `project_helper` module contains utility functions and graph functions. The `project_tests` contains the unit tests for all the problems.

### Install Packages

In [ ]:
import sys
!{sys.executable} -m pip install -r requirements.txt

    100% |████████████████████████████████| 880kB 10.7MB/s ta 0:00:01
    100% |████████████████████████████████| 17.0MB 2.0MB/s eta 0:00:01  2% |▉                               | 419kB 20.9MB/s eta 0:00:01    19% |██████▎                         | 3.3MB 20.2MB/s eta 0:00:01    25% |████████▏                       | 4.3MB 22.3MB/s eta 0:00:01    43% |██████████████                  | 7.5MB 22.2MB/s eta 0:00:01    66% |█████████████████████▎          | 11.3MB 20.3MB/s eta 0:00:01    72% |███████████████████████         | 12.2MB 21.8MB/s eta 0:00:01    77% |████████████████████████▊       | 13.1MB 19.2MB/s eta 0:00:01    82% |██████████████████████████▌     | 14.1MB 20.5MB/s eta 0:00:01    94% |██████████████████████████████▎ | 16.1MB 19.7MB/s eta 0:00:01
    100% |████████████████████████████████| 26.2MB 1.7MB/s eta 0:00:01   11% |███▋                            | 3.0MB 20.3MB/s eta 0:00:02    15% |████▉                           | 4.0MB 20.9MB/s eta 0:00:02    19% |██████              

    100% |████████████████████████████████| 102kB 8.6MB/s a 0:00:01
    100% |████████████████████████████████| 71kB 7.9MB/s ta 0:00:01
    100% |████████████████████████████████| 81kB 5.9MB/s ta 0:00:01    50% |████████████████▏               | 40kB 9.3MB/s eta 0:00:01
  Running setup.py bdist_wheel for cvxpy ... \

### Load Packages

In [2]:
import pandas as pd
import numpy as np
import helper
import project_helper
import project_tests

## Market Data
### Load Data
The data we use for most of the projects is end of day data. This contains data for many stocks, but we'll be looking at stocks in the S&P 500. We also made things a little easier to run by narrowing down our range of time period instead of using all of the data.

In [56]:
df = pd.read_csv('../../data/project_1/eod-quotemedia.csv', parse_dates=['date'], index_col=False)

close = df.reset_index().pivot(index='date', columns='ticker', values='adj_close')

print('Loaded Data')

Loaded Data


### View Data
Run the cell below to see what the data looks like for `close`.

In [1]:
project_helper.print_dataframe(close)

NameError: name 'project_helper' is not defined

### Stock Example
Let's see what a single stock looks like from the closing prices. For this example and future display examples in this project, we'll use Apple's stock (AAPL). If we tried to graph all the stocks, it would be too much information.

In [2]:
apple_ticker = 'AAPL'
project_helper.plot_stock(close[apple_ticker], '{} Stock'.format(apple_ticker))

NameError: name 'project_helper' is not defined

## Resample Adjusted Prices

The trading signal you'll develop in this project does not need to be based on daily prices, for instance, you can use month-end prices to perform trading once a month. To do this, you must first resample the daily adjusted closing prices into monthly buckets, and select the last observation of each month.

Implement the `resample_prices` to resample `close_prices` at the sampling frequency of `freq`.

In [3]:
def resample_prices(close_prices, freq='M'):
    """
    Resample close prices for each ticker at specified frequency.
    
    Parameters
    ----------
    close_prices : DataFrame
        Close prices for each ticker and date
    freq : str
        What frequency to sample at
        For valid freq choices, see http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases
    
    Returns
    -------
    prices_resampled : DataFrame
        Resampled prices for each ticker and date
    """
    # TODO: Implement Function
    resample=close_prices.resample(freq).last()
    return resample
   

project_tests.test_resample_prices(resample_prices)

NameError: name 'project_tests' is not defined

### View Data
Let's apply this function to `close` and view the results.

In [ ]:
monthly_close = resample_prices(close)

#print(monthly_close)
project_helper.plot_resampled_prices(
    monthly_close.loc[:, apple_ticker],
    close.loc[:, apple_ticker],
    '{} Stock - Close Vs Monthly Close'.format(apple_ticker))

## Compute Log Returns

Compute log returns ($R_t$) from prices ($P_t$) as your primary momentum indicator:

$$R_t = log_e(P_t) - log_e(P_{t-1})$$

Implement the `compute_log_returns` function below, such that it accepts a dataframe (like one returned by `resample_prices`), and produces a similar dataframe of log returns. Use Numpy's [log function](https://docs.scipy.org/doc/numpy/reference/generated/numpy.log.html) to help you calculate the log returns.

In [60]:
def compute_log_returns(prices):
    """
    Compute log returns for each ticker.
    
    Parameters
    ----------
    prices : DataFrame
        Prices for each ticker and date
    
    Returns
    -------
    log_returns : DataFrame
        Log returns for each ticker and date
    """
    # TODO: Implement Function
    shifted_prices = prices.shift(1)
    computed_returns = np.log(prices)-np.log(shifted_prices)
     
    
    return computed_returns

project_tests.test_compute_log_returns(compute_log_returns)

Tests Passed


### View Data
Using the same data returned from `resample_prices`, we'll generate the log returns.

In [ ]:
monthly_close_returns = compute_log_returns(monthly_close)
project_helper.plot_returns(
    monthly_close_returns.loc[:, apple_ticker],
    'Log Returns of {} Stock (Monthly)'.format(apple_ticker))

## Shift Returns
Implement the `shift_returns` function to shift the log returns to the previous or future returns in the time series. For example, the parameter `shift_n` is 2 and `returns` is the following:

```
                           Returns
               A         B         C         D
2013-07-08     0.015     0.082     0.096     0.020     ...
2013-07-09     0.037     0.095     0.027     0.063     ...
2013-07-10     0.094     0.001     0.093     0.019     ...
2013-07-11     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...
```

the output of the `shift_returns` function would be:
```
                        Shift Returns
               A         B         C         D
2013-07-08     NaN       NaN       NaN       NaN       ...
2013-07-09     NaN       NaN       NaN       NaN       ...
2013-07-10     0.015     0.082     0.096     0.020     ...
2013-07-11     0.037     0.095     0.027     0.063     ...
...            ...       ...       ...       ...
```
Using the same `returns` data as above, the `shift_returns` function should generate the following with `shift_n` as -2:
```
                        Shift Returns
               A         B         C         D
2013-07-08     0.094     0.001     0.093     0.019     ...
2013-07-09     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...       ...
...            ...       ...       ...       ...       ...
...            NaN       NaN       NaN       NaN       ...
...            NaN       NaN       NaN       NaN       ...
```
_Note: The "..." represents data points we're not showing._

In [64]:
def shift_returns(returns, shift_n):
    """
    Generate shifted returns
    
    Parameters
    ----------
    returns : DataFrame
        Returns for each ticker and date
    shift_n : int
        Number of periods to move, can be positive or negative
    
    Returns
    -------
    shifted_returns : DataFrame
        Shifted returns for each ticker and date
    """
    # TODO: Implement Function
    
    return returns.shift(shift_n)

project_tests.test_shift_returns(shift_returns)

Tests Passed


### View Data
Let's get the previous month's and next month's returns.

In [65]:
prev_returns = shift_returns(monthly_close_returns, 1)
lookahead_returns = shift_returns(monthly_close_returns, -1)

project_helper.plot_shifted_returns(
    prev_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Previous Returns of {} Stock'.format(apple_ticker))
project_helper.plot_shifted_returns(
    lookahead_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Lookahead Returns of {} Stock'.format(apple_ticker))

## Generate Trading Signal

A trading signal is a sequence of trading actions, or results that can be used to take trading actions. A common form is to produce a "long" and "short" portfolio of stocks on each date (e.g. end of each month, or whatever frequency you desire to trade at). This signal can be interpreted as rebalancing your portfolio on each of those dates, entering long ("buy") and short ("sell") positions as indicated.

Here's a strategy that we will try:
> For each month-end observation period, rank the stocks by _previous_ returns, from the highest to the lowest. Select the top performing stocks for the long portfolio, and the bottom performing stocks for the short portfolio.

Implement the `get_top_n` function to get the top performing stock for each month. Get the top performing stocks from `prev_returns` by assigning them a value of 1. For all other stocks, give them a value of 0. For example, using the following `prev_returns`:

```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0.015     0.082     0.096     0.020     0.075     0.043     0.074
2013-07-09     0.037     0.095     0.027     0.063     0.024     0.086     0.025
...            ...       ...       ...       ...       ...       ...       ...
```

The function `get_top_n` with `top_n` set to 3 should return the following:
```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0         1         1         0         1         0         0
2013-07-09     0         1         0         1         0         1         0
...            ...       ...       ...       ...       ...       ...       ...
```
*Note: You may have to use Panda's [`DataFrame.iterrows`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.iterrows.html) with [`Series.nlargest`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.Series.nlargest.html) in order to implement the function. This is one of those cases where creating a vecorization solution is too difficult.*

In [30]:
def get_top_n(prev_returns, top_n):
    """
    Select the top performing stocks
    
    Parameters
    ----------
    prev_returns : DataFrame
        Previous shifted returns for each ticker and date
    top_n : int
        The number of top performing stocks to get
    
    Returns
    -------
    top_stocks : DataFrame
        Top stocks for each ticker and date marked with a 1
    """
    # TODO: Implement Function
    '''
    print(prev_returns)
    
    top_stocks = pd.DataFrame(0,columns = prev_returns.columns, index = prev_returns.index)
    
    df = prev_returns.apply(lambda x: x.nlargest(top_n),axis=1)
    
    top_stocks = top_stocks.merge(df, how = 'right').set_index(df.index)
    #print (top_stocks)
    #print (df)
    top_stocks = (top_stocks.notnull())*1
    top_stocks.astype(np.int64)
    print (top_stocks)
    return top_stocks
   
    print(prev_returns)
    for row in prev_returns.iterrows():
        series = pd.Series(row)
        if series.isnull ==False :
            print(series)
            temp_row = series
            top_row = temp_row.nlargest(top_n) '''
    
    #print (prev_returns.series.dtype)   
    top_returns = prev_returns.fillna(0)
    print (prev_returns) 
    print(top_n)
    
    for i, row in top_returns.iterrows():
        #print(f"Index: {i}")
        #print(f"{row}\n")
        print(top_returns.loc[i].nlargest(top_n))
        top_returns.loc[i]=top_returns.loc[i].nlargest(top_n)
    top_returns = top_returns.fillna(0) 
    print(top_returns)  
    top_returns = top_returns.astype(bool)
    top_returns = top_returns.astype(int)
    print(top_returns)    
    
    return top_returns 


project_tests.test_get_top_n(get_top_n)

                   PCL        NHNU         HMY         QBY        VGBN
2008-08-31         nan         nan         nan         nan         nan
2008-09-30         nan         nan         nan         nan         nan
2008-10-31  3.13172138  0.72709204  5.76874778  1.77557845  0.04098317
2008-11-30 -3.78816218 -0.67583590 -4.95433863 -1.67093250 -0.24929051
3
PCL    0.00000000
NHNU   0.00000000
HMY    0.00000000
Name: 2008-08-31 00:00:00, dtype: float64
PCL    0.00000000
NHNU   0.00000000
HMY    0.00000000
Name: 2008-09-30 00:00:00, dtype: float64
HMY   5.76874778
PCL   3.13172138
QBY   1.77557845
Name: 2008-10-31 00:00:00, dtype: float64
VGBN   -0.24929051
NHNU   -0.67583590
QBY    -1.67093250
Name: 2008-11-30 00:00:00, dtype: float64
                  PCL        NHNU        HMY         QBY        VGBN
2008-08-31 0.00000000  0.00000000 0.00000000  0.00000000  0.00000000
2008-09-30 0.00000000  0.00000000 0.00000000  0.00000000  0.00000000
2008-10-31 3.13172138  0.00000000 5.76874778  1.7755

### View Data
We want to get the best performing and worst performing stocks. To get the best performing stocks, we'll use the `get_top_n` function. To get the worst performing stocks, we'll also use the `get_top_n` function. However, we pass in `-1*prev_returns` instead of just `prev_returns`. Multiplying by negative one will flip all the positive returns to negative and negative returns to positive. Thus, it will return the worst performing stocks.

In [31]:
top_bottom_n = 50
df_long = get_top_n(prev_returns, top_bottom_n)
df_short = get_top_n(-1*prev_returns, top_bottom_n)
project_helper.print_top(df_long, 'Longed Stocks')
project_helper.print_top(df_short, 'Shorted Stocks')

NameError: name 'prev_returns' is not defined

## Projected Returns
It's now time to check if your trading signal has the potential to become profitable!

We'll start by computing the net returns this portfolio would return. For simplicity, we'll assume every stock gets an equal dollar amount of investment. This makes it easier to compute a portfolio's returns as the simple arithmetic average of the individual stock returns.

Implement the `portfolio_returns` function to compute the expected portfolio returns. Using `df_long` to indicate which stocks to long and `df_short` to indicate which stocks to short, calculate the returns using `lookahead_returns`. To help with calculation, we've provided you with `n_stocks` as the number of stocks we're investing in a single period.

In [48]:
def portfolio_returns(df_long, df_short, lookahead_returns, n_stocks):
    """
    Compute expected returns for the portfolio, assuming equal investment in each long/short stock.
    
    Parameters
    ----------
    df_long : DataFrame
        Top stocks for each ticker and date marked with a 1
    df_short : DataFrame
        Bottom stocks for each ticker and date marked with a 1
    lookahead_returns : DataFrame
        Lookahead returns for each ticker and date
    n_stocks: int
        The number number of stocks chosen for each month
    
    Returns
    -------
    portfolio_returns : DataFrame
        Expected portfolio returns for each ticker and date
    """
    # TODO: Implement Function
    print(df_long)
    print(df_short)
    print(lookahead_returns)
    print(n_stocks)
    
    print(df_long*lookahead_returns/n_stocks)
    print(df_short*lookahead_returns/n_stocks)
    
    return (df_long*lookahead_returns/n_stocks) - (df_short*lookahead_returns/n_stocks)

project_tests.test_portfolio_returns(portfolio_returns)

            CFIV  QLOP  CTRG  CSY  IPQ
2008-08-31     0     0     0    0    0
2008-09-30     0     0     0    0    0
2008-10-31     1     0     1    1    0
2008-11-30     0     1     0    1    1
            CFIV  QLOP  CTRG  CSY  IPQ
2008-08-31     0     0     0    0    0
2008-09-30     0     0     0    0    0
2008-10-31     0     1     0    1    1
2008-11-30     1     1     1    0    0
                  CFIV        QLOP        CTRG         CSY         IPQ
2008-08-31  3.13172138  0.72709204  5.76874778  1.77557845  0.04098317
2008-09-30 -3.78816218 -0.67583590 -4.95433863 -1.67093250 -0.24929051
2008-10-31  0.05579709  0.29199789  0.00697116  1.05956179  0.30686995
2008-11-30  1.25459098  6.87369275  2.58265839  6.92676837  0.84632677
3
                  CFIV        QLOP        CTRG         CSY         IPQ
2008-08-31  0.00000000  0.00000000  0.00000000  0.00000000  0.00000000
2008-09-30 -0.00000000 -0.00000000 -0.00000000 -0.00000000 -0.00000000
2008-10-31  0.01859903  0.00000000  0.00

### View Data
Time to see how the portfolio did.

In [49]:
expected_portfolio_returns = portfolio_returns(df_long, df_short, lookahead_returns, 2*top_bottom_n)
project_helper.plot_returns(expected_portfolio_returns.T.sum(), 'Portfolio Returns')

NameError: name 'df_long' is not defined

## Statistical Tests
### Annualized Rate of Return

In [50]:
expected_portfolio_returns_by_date = expected_portfolio_returns.T.sum().dropna()
portfolio_ret_mean = expected_portfolio_returns_by_date.mean()
portfolio_ret_ste = expected_portfolio_returns_by_date.sem()
portfolio_ret_annual_rate = (np.exp(portfolio_ret_mean * 12) - 1) * 100

print("""
Mean:                       {:.6f}
Standard Error:             {:.6f}
Annualized Rate of Return:  {:.2f}%
""".format(portfolio_ret_mean, portfolio_ret_ste, portfolio_ret_annual_rate))

NameError: name 'expected_portfolio_returns' is not defined

The annualized rate of return allows you to compare the rate of return from this strategy to other quoted rates of return, which are usually quoted on an annual basis. 

### T-Test
Our null hypothesis ($H_0$) is that the actual mean return from the signal is zero. We'll perform a one-sample, one-sided t-test on the observed mean return, to see if we can reject $H_0$.

We'll need to first compute the t-statistic, and then find its corresponding p-value. The p-value will indicate the probability of observing a mean return equally or more extreme than the one we observed if the null hypothesis were true. A small p-value means that the chance of observing the mean we observed under the null hypothesis is small, and thus casts doubt on the null hypothesis. It's good practice to set a desired level of significance or alpha ($\alpha$) _before_ computing the p-value, and then reject the null hypothesis if $p < \alpha$.

For this project, we'll use $\alpha = 0.05$, since it's a common value to use.

Implement the `analyze_alpha` function to perform a t-test on the sample of portfolio returns. We've imported the `scipy.stats` module for you to perform the t-test.

Note: [`scipy.stats.ttest_1samp`](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.stats.ttest_1samp.html) performs a two-sided test, so divide the p-value by 2 to get 1-sided p-value

In [76]:
from scipy import stats

def analyze_alpha(expected_portfolio_returns_by_date):
    """
    Perform a t-test with the null hypothesis being that the expected mean return is zero.
    
    Parameters
    ----------
    expected_portfolio_returns_by_date : Pandas Series
        Expected portfolio returns for each date
    
    Returns
    -------
    t_value
        T-statistic from t-test
    p_value
        Corresponding p-value
    """
    # TODO: Implement Function
    print(expected_portfolio_returns_by_date)
    ttest=stats.ttest_1samp(expected_portfolio_returns_by_date,0.0)
    p_value=ttest[1]
    print(ttest)
    
    return ttest[0],ttest[1]/2

project_tests.test_analyze_alpha(analyze_alpha)

2008-08-31    0.00000000
2008-09-30    0.00000000
2008-10-31    0.01859903
2008-11-30   -0.41819699
dtype: float64
Ttest_1sampResult(statistic=-0.94076445661840524, pvalue=0.41622819641354447)
Tests Passed


### View Data
Let's see what values we get with our portfolio. After you run this, make sure to answer the question below.

In [77]:
t_value, p_value = analyze_alpha(expected_portfolio_returns_by_date)
print("""
Alpha analysis:
 t-value:        {:.3f}
 p-value:        {:.6f}
""".format(t_value, p_value))

NameError: name 'expected_portfolio_returns_by_date' is not defined

### Question: What p-value did you observe? And what does that indicate about your signal?

*#TODO: Put Answer In this Cell*
my p-value is 0.41622819641354447 and the alpha value is 0.05 since my p-value is greater than the alpha , the null hypothesis cannot be rejected and the signal is not highly accurate

## Submission
Now that you're done with the project, it's time to submit it. Click the submit button in the bottom right. One of our reviewers will give you feedback on your project with a pass or not passed grade. You can continue to the next section while you wait for feedback.